### PR TITLE
Fix #3441: parentheses wrapping expression throw invalid error 

### LINKED ISSUE #3441: Lexer: Parentheses wrapping expression throw invalid error in `then` clause
Why is this valid

``` coffeescript
(b(); break) for i in j
```

But this is not?

``` coffeescript
a = -> (b(); break) for i in j
```

### PR DESCRIPTION
**Pull request by zdenko**
Fixes #3441.

Cases where `statement` is used inside parentheses don't compile:
```coffeescript
(a(); break) for a in obj
# error: cannot use a pure statement in an expression
# (a(); break) for a in obj
#       ^^^^^

(a(); break) while a
# error: cannot use a pure statement in an expression
# (a(); break) while a
#       ^^^^^

for a in obj then (a(); break)
# error: cannot use a pure statement in an expression
# for a in obj then (a(); break)
#                         ^^^^^

while a then (a(); break)
# error: cannot use a pure statement in an expression
# while a then (a(); break)
#                    ^^^^^
```

With this PR:
```coffeescript

# (a(); break) for a in obj
# for a in obj then (a(); break)
for (i = 0, len = obj.length; i < len; i++) {
  a = obj[i];
  a();
  break;
}

# (a(); break) while a
# while a then (a(); break)
while (a) {
  a();
  break;
}
```

When `Parens` block includes a `StatementLiteral` (e.g. `(b; break) for a in arr`), it won't compile since `Parens` (`(b; break)`) is compiled as `Value` and pure statement (`break`) can't be used in an expression.  

This PR checks `Parenthetical` rule in the grammar and returns `Block` instead of `Parens`  if `StatementLiteral` is detected. The `Block` node is passed further to the `Expression`.

### CODE REVIEW

File: src/grammar.coffee [hunk 1]
```diff
@@ -597,8 +601,22 @@ grammar =
...
   # the trick.
   Parenthetical: [
-    o '( Body )',                               -> new Parens $2
-    o '( INDENT Body OUTDENT )',                -> new Parens $3
+    o '( Body )',                               ->
+        # When `Parens` block includes a `StatementLiteral` (e.g. `(b; break) for a in arr`),
+        # it won't compile since `Parens` (`(b; break)`) is compiled as `Value` and
+        # pure statement (`break`) can't be used in an expression.
+        # For this reasons, we return `Block`, which is passed further to `Value` and `Expression`.
+        stm = $2.contains (n) -> n.constructor.name is 'StatementLiteral'
```
Comment (vendethiel) [inline src/grammar.coffee]: I'd rather we don't add such code to the grammar. We can have a node function for that.

Comment (vendethiel) [reply src/grammar.coffee]: Also we try not to use `constructor.name` for obfuscation reasons IIRC.

File: src/grammar.coffee [hunk 2]
```diff
@@ -334,7 +334,7 @@ grammar =
     o 'Assignable'
     o 'Literal',                                -> new Value $1
     o 'Parenthetical',                          ->
-        if $1.constructor.name is 'Parens'
+        if $1.hasParentheses()
```
Comment (vendethiel) [inline src/grammar.coffee]: you can use `unwrap`

Comment (zdenko) [reply src/grammar.coffee]: Nice.

**Pull request merged**